ngraph/core/src/op/matmul.cpp

```diff
@@ -113,68 +53,170 @@ shared_ptr<Node> op::MatMul::clone_with_new_inputs(const OutputVector& new_args)
+        std::vector<Dimension> output_shape(max_rank);
+
+        // Handle batch size
+        if (max_rank > 2)
         {
-            NGRAPH_CHECK(arg0_shape_update[arg0_rank - 1] == arg1_shape_update[arg1_rank - 2],
-                         "Incompatible arg shapes");
+            std::vector<Dimension> small_batch_matrix =
+                arg0_rank > arg1_rank ? arg1_shape_tmp : arg0_shape_tmp;
+            std::vector<Dimension> big_batch_matrix =
+                arg0_rank > arg1_rank ? arg0_shape_tmp : arg1_shape_tmp;
+
+            if (arg0_rank != arg1_rank)
+            {
+                // Expand small_batch_matrix (with 1) to have the same rank as big_batch_matrix
+                size_t delta_rank = big_batch_matrix.size() - small_batch_matrix.size();
+                small_batch_matrix.insert(small_batch_matrix.begin(), delta_rank, 1);
+            }
+
+            // Broadcast all batches (max_rank - 2),
+            // expand dim with value 1 to bigger dim if dimensions are not equal
+            for (auto i = 0; i < max_rank - 2; i++)
+            {
+                // Both dimensions are static
+                if (small_batch_matrix[i].is_static() && big_batch_matrix[i].is_static())
+                {
+                    // Check if batch dimension broadcast is possible
+                    NGRAPH_CHECK(
+                        (small_batch_matrix[i].get_length() == big_batch_matrix[i].get_length()) ||
+                            (std::min(small_batch_matrix[i].get_length(),
+                                      big_batch_matrix[i].get_length()) == 1),
+                        "Incompatible batch dimensions.");
 
-            const auto& broadcast_shapes = builder::get_numpy_broadcast_shapes(
-                {Shape{begin(arg0_shape_update), next(end(arg0_shape_update), -2)},
-                 Shape{begin(arg1_shape_update), next(end(arg1_shape_update), -2)}});
+                    output_shape[i] = std::max(small_batch_matrix[i].get_length(),
+                                               big_batch_matrix[i].get_length());
+                }
+                // At least one dimension is dynamic
+                else
+                {
+                    // Dimension with value 1 can be expanded to any bigger
+                    auto min_dim_val = std::min(small_batch_matrix[i].get_min_length(),
+                                                big_batch_matrix[i].get_min_length());
+                    if (min_dim_val > 1)
+                    {
+                        NGRAPH_CHECK(Dimension::merge(merged_dimension,
+                                                      small_batch_matrix[i],
+                                                      big_batch_matrix[i]),
+                                     "Incompatible batch dimensions bounds.");
+                        output_shape[i] = merged_dimension;
+                        continue;
+                    }
```

Comment:
I think that you can produce shorter code by not differentiating static and dynamic cases here and below. The current code is huge and hard to follow and maintain. Dimension::merge works fine for static shapes as well. Just check if 1 is included into a range and proceed accordingly avoid explicit touching get_length attribute for a dimension. Does it make sense, @mitruska? Could you try?

Reply:
Thank you for the suggestion, I refactored and reduced this part of code.